### ISSUE TITLE
Request to add check for math injection

### ISSUE THREAD

**Issue by bnevis-i** [koalaman/shellcheck#2871]
#### For new checks and feature suggestions
- [X] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

This test case is inspired by <https://research.nccgroup.com/2020/05/12/shell-arithmetic-expansion-and-evaluation-abuse/>

```sh
#!/bin/bash
test -f counter || echo 1 > counter
COUNTER=$(cat counter)
[[ -z $COUNTER ]] && exit 1
(( COUNTER++ ))
printf "%s" "$COUNTER" > counter
echo "$COUNTER"
echo Done
```

```shell
$./shellinject.sh
2
Done

$ echo '_[$(ls)]' > counter
$./shellinject.sh
(directory listing)
```

#### Here's what shellcheck currently says:

```shell
$ shellcheck shellinject.sh 
$
```

#### Here's what I wanted or expected to see:

Warning about math injection.

**Comment by gsauthof** [comment]
I agree that injection via arithmetic evaluation is a serious foot gun and it would be great if shellcheck would warn about it.

However, your test case doesn't work as advertised:

```
[juser@localhost ~]$ echo '_[$(ls)]' > counter
[juser@localhost ~]$ ./shellinject.sh
./shellinject.sh: line 5: counter
shellinject.sh: syntax error: invalid arithmetic operator (error token is ".sh")
_[$(ls)]
Done
```

The thing is that stdout of that sub-shell is captured and thus you don't see the ls output.

---

A minimal example that demonstrates the issue in a more obvious way:

```
[juser@localhost ~]$ cat > foo.sh
if [[ "$1" -gt 23 ]]; then echo foo; fi
[juser@localhost ~]$ cat > bar.sh
echo say what >&2
[juser@localhost ~]$ chmod 755 bar.sh
[juser@localhost ~]$ bash foo.sh 'a[$(./bar.sh)]'
say what
```

I can confirm that right now https://www.shellcheck.net/ does **not** warn about this foo.sh snippet.

---

See also some [additional sources](https://gms.tf/inject_code.html#related-work) on this topic that predate the nccgroup link several years, in the related work section of a blog post I authored.

**Comment by wileyhy** [comment]
It would be a good get for sure.
